src/V3EmitMk.cpp

```diff
@@ -31,11 +31,557 @@ class EmitMk final {
+            if (ofIt->isConcatenatingFile()) {
+                for (const auto& ocf : ofIt->concatenatedFileNames) {
+                    UASSERT(ifIt != inputFiles.end(),
+                            "More output files than input files. First extra file: " << ocf);
+                    UASSERT(ifIt->name == ocf, "Name mismatch: " << ifIt->name << " != " << ocf);
+                    ++ifIt;
+                }
+                ++ofIt;
+            } else {
+                UASSERT(ifIt->name == ofIt->fileName,
+                        "Name mismatch: " << ifIt->name << " != " << ofIt->fileName);
+                ++ifIt;
+                ++ofIt;
+            }
+        }
+        UASSERT(ifIt == inputFiles.end(),
+                "More input files than input files. First extra file: " << ifIt->name);
+        UASSERT(ofIt == outputFiles.end(),
+                "More output files than input files. First extra file: " << ofIt->fileName);
+    }
+
+    static std::vector<FileOrConcatenatedFilesList>
+    singleConcatenatedFilesList(std::vector<FileNameWithScore> inputFiles, std::int64_t totalScore,
+                                std::string concatenatingFilePrefix) {
+        UINFO(4, __FUNCTION__ << ":" << endl);
+        UINFO(5, "Number of input files: " << inputFiles.size() << endl);
+        UINFO(5, "Total score: " << totalScore << endl);
+        UINFO(5, "Concatenated file prefix: " << concatenatingFilePrefix << endl);
+
+        // Called when the concatenation is aborted
+        static const auto convertInputFilesToOutputFiles
+            = [](std::vector<FileNameWithScore> inputFiles) {
+                  std::vector<FileOrConcatenatedFilesList> outputFiles;
+                  outputFiles.reserve(inputFiles.size());
+                  for (auto& file : inputFiles) {
+                      outputFiles.push_back({/*fileName=*/std::move(file.name)});
+                  }
+                  return outputFiles;
+              };
+
+        const int totalBucketsNum = v3Global.opt.outputConcatenationGroups();
+        UINFO(5, "Number of buckets: " << totalBucketsNum << endl);
+        UASSERT(totalBucketsNum > 0, "More than 0 buckets required");
+
+        // MAIN PARAMETERS
+
+        // Minimum number of input files required to perform concatenation.
+        // Concatenation of a small number of files does not give any performance advantages.
+        // The value has been chosen arbitrarily, most likely could be larger.
+        constexpr std::size_t MIN_FILES_COUNT = 100;
+
+        // Concatenation of only a few files most likely does not increase performance.
+        // The value has been chosen arbitrarily.
+        constexpr std::size_t MIN_FILES_PER_BUCKET = 4;
+
+        // ---
+
+        {
+            // Return early if there's nothing to do.
+            bool returnEarly = false;
+            if (inputFiles.size() < MIN_FILES_COUNT) {
+                UINFO(0, "File concatenation skipped: Too few files ("
```

Comment:
Don't have any UINFO 0's.  This seems something only to print when this feature is debugged, so 4 seems right.